### PR TITLE
fix: Add defaults field values before validating schema

### LINKED ISSUE #4655: Wrong schema validation
**I'm submitting a ...**  (check one with "x")
- [x] bug report
- [ ] feature request
- [ ] support request => Please do not submit support requests here, instead ask your query in out Gitter channel at https://gitter.im/fossasia/open-event-server

**Current behavior:**
Schema validation on the server is broken

```JSON
{
  "data": {
    "relationships": {
      "event": {
        "data": {
          "type": "event",
          "id": "{{event}}"
        }
      }
    },
    "attributes": {
      "name": "ticket-name",
      "description": "some description",
      "type": "VIP-ticket",
      "price": "100.00",
      "quantity": "100",
      "is-description-visible": null,
      "position": "1",
      "is-fee-absorbed": "false",
      "sales-starts-at": "2017-06-01T01:24:47.500127+00:00",
      "sales-ends-at": "2017-06-30T00:24:47.500127+00:00",
      "is-hidden": "false",
      "min-order": "1",
      "max-order": "100"
    },
    "type": "ticket"
  }
}
```
If any of the optional field is null (which is what it can be when serializing in Java or any other language with schema, it gives the following error

```JSON
{
    "jsonapi": {
        "version": "1.0"
    },
    "errors": [
        {
            "status": "422",
            "source": {
                "pointer": "/data/attributes/is-description-visible"
            },
            "title": "Validation error",
            "detail": "Field may not be null."
        }
    ]
}
```

However this is clearly an optional field. Also, all validation is done prematurely and before applying the default values, which is counter productive. If quantity is missing, it should be assigned the default value from the server rather than reporting `quantity should be higher than min-order`

Some of these changes in API are breaking and make it really hard developing mobile applications, because we cannot base our development on a particular API assuming it won't break any contracts. master branch is unusable so we switched to development, but it introduces several breaking changes without any communication in documentation.

Several times, the development in mobile (and I assume web) is halted with questions as we are not able to POST ticket, feedback, event because of some reason, and it prevents us to create new features or move forward in a direction we believe API will move. But we cannot because it is unpredictable what will change and how will we accomodate with it. Now we are essentially stopped in our tracks until the server is fixed because not everyone knows how to setup API locally and even if they do, they don't know which branch or commit they should target? And it might be easy to develop with local API on web, it isn't on mobile

~~I believe this is a side effect of the recent change in flask rest jsonapi, and the changes made in local fork might be necessary and convenient but they break several previously established behaviours~~. I think we should agree that APIs should conform to their users and not the other way around, so I urge the developers to ensure that any such breaking changes be well tested before merging and communicated in some form of documentation. Thanks

**Edit:** The dev branch was not deployed for a long time to heroku, so all the breaking changes throughout the months got introduced to us in last few weeks.

### PR DESCRIPTION
**Pull request by srv-twry**
Fixes #4655 

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [ ] The unit tests pass locally with my changes 
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

- [x] All the functions created/modified in this PR contain relevant docstrings.

#### Short description of what this resolves:
Fixes the wrong schema validation. Default values are now added to the data before deserializing it.

#### Changes proposed in this pull request:

- Add default values to the fields before schema validation

**TODO**:

- [x] Add the decorator to all the schema classes
- [x] Write a few tests
- [x] Add necessary docstrings

~~~@iamareebjamal I have opened the WIP PR so as to get feedback on the way the decorator has been implemented. Is it fine ? Moreover suggestions of the name of the decorator is also welcomed.~~~

### CODE REVIEW

**Comment by srv-twry** [comment]
@iamareebjamal Ping.

**Comment by srv-twry** [comment]
@niranjan94 @shubham-padia 
Please review so that i can add the decorator to all the models. 
For the context, please refer to the discussion in the issue.

**Comment by srv-twry** [comment]
@niranjan94 @iamareebjamal Ready for review

File: tests/unittests/utils/test_common.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+from pytz import timezone
+
+from app import current_app as app
+from app.api.schema.tickets import TicketSchema
+from tests.unittests.setup_database import Setup
+from tests.unittests.utils import OpenEventTestCase
+from utils.common import patch_defaults
+
+
+class TestUtils(OpenEventTestCase):
```
Comment (iamareebjamal) [inline tests/unittests/utils/test_common.py]: If it does not depend on the db, don't use OpenEventTestCase

**Comment by maxlorenz** [comment]
One minor thing: Would it be better to use a new schema for the test only as every schema change in the original can break this (somewhat unrelated) test?

**Comment by iamareebjamal** [comment]
Yup I agree. Test should have a test schema. And a couple more tests actually. To test the output as well

**Comment by srv-twry** [comment]
@iamareebjamal @maxlorenz Please review again. I have a separate TestSchema and tests using it.

File: tests/unittests/test_schema.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+from marshmallow_jsonapi import fields
+from marshmallow_jsonapi.flask import Schema
+
+from app.api.helpers.utilities import dasherize
+
+
+class TestSchema(Schema):
```
Comment (iamareebjamal) [inline tests/unittests/test_schema.py]: Make a file `test_schema_defaulting` and move both the class and tests in it

**Comment by srv-twry** [comment]
@iamareebjamal Please suggest changes when you are free :)

File: app/api/schema/custom_forms.py [hunk 3]
```diff
@@ -3,8 +3,10 @@
 from marshmallow_jsonapi.flask import Schema, Relationship
 
 from app.api.helpers.utilities import dasherize
+from utils.common import use_defaults
 
 
+@use_defaults()
```
Comment (iamareebjamal) [inline app/api/schema/custom_forms.py]: Looks good. Remove paranthesis, as they are not needed

Comment (srv-twry) [reply app/api/schema/custom_forms.py]: It need ```self``` right ?

**Comment by srv-twry** [comment]
@niranjan94 Please review and merge at your comfort :)

**Pull request merged**